classifier-comparision/evaluate.py

```diff
@@ -0,0 +1,167 @@
+        "hv": defaultdict(list),
+        "human": defaultdict(list),
+    }
+    cdata["processed"] = {}
+
+    all_reads = set()
+
+    with open(os.path.join(cdata["dashboard_dir"],
+                           "hvreads",
+                           "%s.hvreads.json" % sample)) as inf:
+        hvr = json.load(inf)
+        for read_id in hvr:
+            cdata["hvreads"][read_id] = hvr[read_id]
+            all_reads.add(read_id)
+
+    for bowtie_db in cdata["alignments"]:
+        with gzip.open(os.path.join(cdata["dashboard_dir"],
+                                    "alignments",
+                                    "%s.%s.alignments.tsv.gz" % (
+                                        sample, bowtie_db)),
+                       "rt") as inf:
+            for line in inf:
+                bits = line.removesuffix("\n").split("\t")
+                read_id = bits[0]
+                all_reads.add(read_id)
+                cdata["alignments"][bowtie_db][read_id].append(bits)
+
+    # This one is so big we process it streaming.
+    for fname in glob.glob(
+            os.path.join(cdata["dashboard_dir"],
+                         "processed", "%s.*.kraken2.tsv.gz" % sample)):
+        with gzip.open(fname, "rt") as inf:
+            for line in inf:
+                bits = line.removesuffix("\n").split("\t")
+                read_id = bits[1]
+                if read_id not in all_reads:
+                    # All of the classifiers need at least an HVR or alignment
+                    # hit, so we can speed things up a lot by ignoring any
+                    # reads that don't have one of these.
+                    continue
+                cdata["processed"] = bits
+                yield read_id, cdata
+
+def read_id_to_sample_id(read_id):
+    return read_id.removeprefix("M_").split(".")[0]
+
+def start(*classifier_names):
+    classifiers = [
+        importlib.import_module(classifier_name)
+        for classifier_name in classifier_names]
+
+    with open("classified-reads.json") as inf:
+        ground_truth = json.load(inf)
+
+    gt_read_ids = set(ground_truth["yes"]) | set(ground_truth["no"])
+    samples = set(read_id_to_sample_id(read_id)
+                  for read_id in gt_read_ids)
+
+    metrics = [Counter() for _ in classifiers]
+    metric_names = [
+        "totpos", "truepos", "trueneg", "falsepos", "falseneg"]
+
+    all_notes = {}
+    for sample in sorted(samples):
+        #if sample != "SRR14530891":
+        #    continue
+        sample_notes = {}
+        for read_id, cdata in load_data(sample):
+            read_notes = {}
+            if read_id in gt_read_ids:
+                read_notes["hv"] = read_id in ground_truth["yes"]
+
+            for classifier_name, classifier, metric in zip(
+                    classifier_names, classifiers, metrics):
+                out = classifier.classify(sample, read_id, cdata)
+                if type(out) == type(True):
+                    result = out
```

Comment:
I don't fully understand how classifier_notes get's handled here, it would be helpful to talk through this.

Reply:
We talked through this in person: notes are debug information that a classifier optionally generates